# Implementation of Natural Language Processing (NLP) using python

Prepared by : Swaminathan Ayyappan

Email : swamynathanayyappan@gmail.com

LinkedIn : linkedin.com/in/swaminathan-ayyappan-60b685175/

Concept of Natural Language Processing :

                        The Natural Language Processing (NLP) is used to make useful insights from the data which consists of text,In the real life examples we use google home , amazon's alexa and apple's siri which basically works under the principle of NLP.
                        
Problem Statement :

                    In France there is a popular newly opened restaurant , people living there have visited to that place and shared their experience of visit in a review ,we are provided with the dataset (Restaurant_Reviews.tsv) our aim is to predict whether the review made by the customer is a positive / negative review   
     
Steps invloved in Natural Language Processing (NLP) :


i) Initially the dataset contains of words are collected and it was subjected to cleaning inorder to remove the special characters and stopwords in it 

ii)Followed by cleaning,bagging of words has been done.

iii)Since our problem statement is to predict of categorical features we can use any kind lf classification algorithms in this problem and here we are going to use RandomForest classification algorithm.


In [1]:
#Importing libraries 
import pandas as pd
import numpy as np

In [2]:
#Importing the dataset into red_data variable with the help of pandas.
res_data=pd.read_csv('F:/STUDY MATERIALS/[FreeCourseSite.com] Udemy - Machine Learning A-Z™ Hands-On Python & R In Data Science/Datasets/Machine Learning A-Z New/Part 7 - Natural Language Processing/Section 36 - Natural Language Processing/Restaurant_Reviews.tsv',delimiter='\t',quoting=3)

In [3]:
#reviewing the first five entries in the res_data dataset.
res_data.head()

Review  Liked
0                           Wow... Loved this place.      1
1                                 Crust is not good.      0
2          Not tasty and the texture was just nasty.      0
3  Stopped by during the late May bank holiday of...      1
4  The selection on the menu was great and so wer...      1

In the above dataset we can see that various set of reviews made by the customers the feature "Liked" indicates whether the review is a positive or a negative review. 

In [4]:
#reviewing the shape of the res_data.
res_data.shape

(1000, 2)

# Cleaning the text

In [5]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
corpus=[]
for i in range (0,1000):
    review=re.sub('[^a-zA-Z]',' ',res_data['Review'][i])
    review=review.lower()
    review=review.split()
    ps=PorterStemmer()
    review=[ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review=' '.join(review)
    corpus.append(review)

# Bagging of words

In [6]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=1500)

## Dependent and Independent variable allocation

In [7]:
X=cv.fit_transform(corpus).toarray()
y=res_data['Liked']

### Train test split

In [8]:
from sklearn.model_selection import train_test_split

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

#### Importing RandomForestClassifier algorithm from ensemble package

In [10]:
from sklearn.ensemble import RandomForestClassifier

In [11]:
rand=RandomForestClassifier(n_estimators=500,random_state=0)

In [12]:
rand.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=500,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [13]:
rand.score(X_test,y_test)*100

76.5

In [14]:
from sklearn.metrics import confusion_matrix,classification_report

In [15]:
confusion_matrix(rand.predict(X_test),y_test)

array([[93, 32],
       [15, 60]], dtype=int64)

In [16]:
classification_report(rand.predict(X_test),y_test)

'              precision    recall  f1-score   support\n\n           0       0.86      0.74      0.80       125\n           1       0.65      0.80      0.72        75\n\n    accuracy                           0.77       200\n   macro avg       0.76      0.77      0.76       200\nweighted avg       0.78      0.77      0.77       200\n'

Thus in the above statement we have an accuracy of 76 % for which is quite good accuracy for our problem statement.